test/common/stats/thread_local_store_test.cc

```diff
@@ -1186,5 +1195,155 @@ TEST_F(HistogramTest, ParentHistogramBucketSummary) {
+      }
+    }
+
+    {
+      BlockingScope blocking_scope(1);
+      main_dispatcher_->post(blocking_scope.run([this]() {
+        tls_ = std::make_unique<ThreadLocal::InstanceImpl>();
+        tls_->registerThread(*main_dispatcher_, true);
+        for (Event::DispatcherPtr& dispatcher : thread_dispatchers_) {
+          // Worker threads must be registered from the main thread, per assert in registerThread().
+          tls_->registerThread(*dispatcher, false);
+        }
+        store_->initializeThreading(*main_dispatcher_, *tls_);
+      }));
+    }
+  }
+
+  ~ClusterShutdownCleanupStarvationTest() {
+    {
+      BlockingScope blocking_scope(1);
+      main_dispatcher_->post(blocking_scope.run([this]() {
+        store_->shutdownThreading();
+        tls_->shutdownGlobalThreading();
+        tls_->shutdownThread();
+      }));
+    }
+
+    for (Event::DispatcherPtr& dispatcher : thread_dispatchers_) {
+      dispatcher->post([&dispatcher]() { dispatcher->exit(); });
+    }
+
+    for (Thread::ThreadPtr& thread : threads_) {
+      thread->join();
+    }
+
+    main_dispatcher_->post([this]() {
+      store_.reset();
+      tls_.reset();
+      main_dispatcher_->exit();
+    });
+    main_thread_->join();
+  }
+
+  void incCounters() {
+    for (uint32_t i = 0; i < NumScopes; ++i) {
+      ScopePtr scope = store_->createScope("scope."); // absl::StrCat("scope", i, "."));
+      scope->counterFromStatName(my_counter_name_).inc();
+    }
+  }
+
+  void workerThreadFn(uint32_t thread_index, BlockingScope& blocking_scope) {
+    thread_dispatchers_[thread_index] = api_->allocateDispatcher();
+    blocking_scope.decrementCount();
+    thread_dispatchers_[thread_index]->run(Event::Dispatcher::RunType::RunUntilExit);
+  }
+
+  void mainThreadFn(BlockingScope& blocking_scope) {
+    main_dispatcher_ = api_->allocateDispatcher();
+    blocking_scope.decrementCount();
+    main_dispatcher_->run(Event::Dispatcher::RunType::RunUntilExit);
+  }
+
+  void incCountersAllThreads() {
+    BlockingScope blocking_scope(NumThreads);
+    for (Event::DispatcherPtr& thread_dispatcher : thread_dispatchers_) {
+      thread_dispatcher->post(blocking_scope.run([this]() { incCounters(); }));
+    }
+  }
+
+  std::chrono::seconds elapsedTime() {
+    return std::chrono::duration_cast<std::chrono::seconds>(time_system_.monotonicTime() -
+                                                            start_time_);
+  }
+
+  Event::TestRealTimeSystem time_system_;
+  MonotonicTime start_time_;
+  Api::ApiPtr api_;
+  Event::DispatcherPtr main_dispatcher_;
+  std::vector<Event::DispatcherPtr> thread_dispatchers_;
+  Thread::ThreadFactory& thread_factory_;
+  std::unique_ptr<ThreadLocal::InstanceImpl> tls_;
+  Thread::ThreadPtr main_thread_;
+  std::vector<Thread::ThreadPtr> threads_;
+  StatNamePool pool_;
+  StatName my_counter_name_;
+};
+
+TEST_F(ClusterShutdownCleanupStarvationTest, EightThreadsWithBlockade) {
```

Comment:
Same comment here with some // comments for each test better describing what we are testing?

Reply:
done